### PR TITLE
Add 2MB option for partitions.csv generation and restore use of user-defined partitions

### LINKED ISSUE #5121: Can no longer use custom partitions table due to the recent fix " Generate partitions.csv based on flash size #5697 "
### The problem

Updating an existing ESP Home ESP32 device with 2MB memory now fails, due to the code now attempting to build a custom partition and ignoring an existing custom partition.

### Which version of ESPHome has the issue?

2023.11.1

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2023.11.2

### What platform are you using?

ESP32-IDF

### Board

esp32-c3-devkitm-1

### Component causing the issue

    esphome/esphome/components/esp32  /__init__.py

### Example YAML snippet

```yaml
esphome:
  name: "water"
  friendly_name: Hot Water
  platformio_options:
    board_upload.flash_size: 2MB
    board_upload.maximum_size: 2097152
    board_build.partitions: "/config/esphome/minimal.csv"
    board_build.flash_mode: dio
#    build_flags: 
#      - "-include \"fix.h\""

esp32:
  board: esp32-c3-devkitm-1
  framework:
    type: esp-idf
    version: recommended
    # Custom sdkconfig options
    sdkconfig_options:
      CONFIG_COMPILER_OPTIMIZATION_SIZE: y
    # Advanced tweaking options
    advanced:
      ignore_efuse_mac_crc: false
```

### Anything in the logs that might be useful for us?

```txt
INFO Generating C++ source...
Traceback (most recent call last):
  File "/usr/local/bin/esphome", line 33, in <module>
    sys.exit(load_entry_point('esphome', 'console_scripts', 'esphome')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/__main__.py", line 1040, in main
    return run_esphome(sys.argv)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/__main__.py", line 1027, in run_esphome
    rc = POST_CONFIG_ACTIONS[args.command](args, config)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/__main__.py", line 457, in command_run
    exit_code = write_cpp(config)
                ^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/__main__.py", line 193, in write_cpp
    return write_cpp_file()
           ^^^^^^^^^^^^^^^^
  File "/esphome/esphome/__main__.py", line 211, in write_cpp_file
    writer.write_cpp(code_s)
  File "/esphome/esphome/writer.py", line 342, in write_cpp
    copy_src_tree()
  File "/esphome/esphome/writer.py", line 295, in copy_src_tree
    copy_files()
  File "/esphome/esphome/components/esp32/__init__.py", line 610, in copy_files
    get_idf_partition_csv(
  File "/esphome/esphome/components/esp32/__init__.py", line 550, in get_idf_partition_csv
    app_partition_size = APP_PARTITION_SIZES[flash_size]
                         ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
KeyError: '2MB'
```

### Additional information

Need to find a way to allow custom partition configurations like we used to have. Also, the auto-partition configuration needs to include support for 2MB ESP-32 boards, like this one:
https://www.waveshare.com/esp-c3-32s-kit.htm

### PR DESCRIPTION
**Pull request by kbx81**
# What does this implement/fix?

Quick fix for https://github.com/esphome/issues/issues/5121

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/issues/issues/5121

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3388

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by jesserockz — APPROVED** [review]

File: esphome/components/esp32/__init__.py [hunk 1]
```diff
@@ -387,20 +416,23 @@ def _detect_variant(value):
...
 CONF_FLASH_SIZE = "flash_size"
+CONF_PARTITIONS = "partitions"
 CONFIG_SCHEMA = cv.All(
     cv.Schema(
         {
             cv.Required(CONF_BOARD): cv.string_strict,
             cv.Optional(CONF_FLASH_SIZE, default="4MB"): cv.one_of(
                 *FLASH_SIZES, upper=True
             ),
+            cv.Optional(CONF_PARTITIONS): cv.file_,
```
Comment (jesserockz) [inline esphome/components/esp32/__init__.py:435]: Actually, should these two be exclusive?

Comment (kbx81) [reply esphome/components/esp32/__init__.py:435]: Hmm, good thought...

Comment (kbx81) [reply esphome/components/esp32/__init__.py:435]: Actually, no, we can't do that as PIO needs to know the flash chip size regardless of the partitions file.

Comment (jesserockz) [reply esphome/components/esp32/__init__.py:435]: Ok good point.

**Pull request merged**